### PR TITLE
docs: add zexuan's nginx + lua doc

### PR DESCRIPTION
**Pull request by yzeng25**
add zexuan's nginx + lua doc

### CODE REVIEW

File: _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
+---
+title: "为什么 Apache APISIX 选择 Nginx + Lua 这个技术栈？"
+avatar: "https://avatars.githubusercontent.com/u/4161644?v=4"
+author: "spacewander"
+href: "https://github.com/spacewander"
+date: 2021-08-25
+keywords: API 网关,APISIX,Lua,Nginx
+description: 本文由深圳支流科技工程师罗泽轩撰写，介绍了 APISIX 选用 Nginx + Lua 这个技术栈的历史背景和这个技术栈为 APISIX 带来的优势。罗泽轩是 OpenResty 开发者以及 Apache APISIX Committer。
```
Comment (Serendipity96) [inline _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md:8]: APISIX -> Apache APISIX

Comment (Serendipity96) [reply _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md:8]: 深圳支流科技 or  api7.ai ?

File: _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+---
+
+笔者在今年的 COSCUP 大会做分享时，曾有观众问这样的问题，为什么 APISIX、Kong 和 3scale 这些网关都采用 Lua 来编写逻辑？
```
Comment (Serendipity96) [inline _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md:11]: Apache APISIX

Comment (Serendipity96) [reply _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md:11]: Maybe the whole article needs to be checked.

File: _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+众所周知，Nginx 是一个 bytes matter 的高性能服务器实现，对内存使用非常抠门。举两个例子：
+
+1. Nginx 里面的 request header 在大多数时候都只是指向原始的 HTTP 请求数据的一个指针，只有在修改的时候才会创建副本。
+2. Nginx 代理上游响应时对 buffer 的复用逻辑非常复杂，是我读过的最为烧脑的代码之一。
+
+凭借这种抠门，Nginx 得以屹立在高性能服务器之巅。
+
+相反的，Go 标准库的 HTTP 实现，是一个滥用内存的典型反例。
+
+这可不是我的一面之辞，Fasthttp，一个重新实现 Go 标准库里面的 HTTP 包的项目，就举了两个例子：
```
Comment (Serendipity96) [inline _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md:41]: Fast HTTP

File: _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+1. 标准库的 HTTP Request 结构体没法复用
+2. headers 总是被提前解析好，存储成 map[string][]string，即使没有用到（原文见：https://github.com/valyala/fasthttp#faq ）
```
Comment (Serendipity96) [inline _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md:44]: What about markdown link？

File: _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+举路由系统作为一个例子，Nginx 的路由需要在配置文件里面进行配置，每次更改路由，都需要 reload 之后才能生效。这是因为 Nginx 的路由分发只支持静态配置，不能动态增减路由。
+
+**为了实现路由动态配置，APISIX 做了两件事：**
+
+1. 在 Nginx 配置文件里面配置单个 server，这个 server 里面只有一个 location。我们把这个 location 作为主入口，这样所有的请求都会走到这个地方上来。
+2. 我们用 Lua 完成路由分发的工作。APISIX 的路由分发模块，支持在运行时增减路由，这样就能动态配置路由了。
+
+你可能会问，在 Lua 里面做路由分发，会比 Nginx 的实现慢吗？
+
+就像前面提到过的一样，凡是对性能要求比较高的，我们会把核心代码用 C 改写。我们的路由分发模块就是这么干的。路由分发模块在匹配路由时，会采用一个前缀树来匹配。而这个前缀树是用 C 实现的。感兴趣的读者可以看下代码：https://github.com/api7/lua-resty-radixtree/ 。
```
Comment (Serendipity96) [inline _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md:130]: markdown link?

File: _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+举这个例子并不是想证明 LuaJIT 比 Go 快 20 倍。我只想说明用 micro benchmark 证明某个语言比另一个语言快的意义不大，因为影响性能的因素很多。一个简单的 micro benchmark 很有可能过分强调某一个因素，导致出乎意料的结果。
+
+## Nginx + Lua ：高性能 + 灵活
+
+让我们转回 APISIX 的 Nginx + Lua 的技术栈。Nginx + Lua 的技术栈给我们带来的，不仅仅是高性能。
+
+经常有人问我们，既然你们是基于 Nginx 开源版本，而 Nginx 并不支持动态配置，为什么 APISIX 声称自己可以实现动态配置？你们是不是改了点东西？
+
+是的，我们确实有在维护自己的 Nginx 发行版，不过 APISIX 的大部分功能在官方的 Nginx 上就能使用。我们之所以能做到动态配置，全靠把配置放到 Lua 代码里面来实现。
```
Comment (Serendipity96) [inline _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md:119]: 维护自己的 Nginx 发行版. double check

File: _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+就像前面提到过的一样，凡是对性能要求比较高的，我们会把核心代码用 C 改写。我们的路由分发模块就是这么干的。路由分发模块在匹配路由时，会采用一个前缀树来匹配。而这个前缀树是用 C 实现的。感兴趣的读者可以看下代码：https://github.com/api7/lua-resty-radixtree/ 。
+
+完成了 C 层面上的前缀树匹配，接下来就该 Lua 发挥灵活性的时刻了。对于匹配同一前缀的各个路由，我们支持通过许多别的方式来进行下一级的匹配，其中就包含通过一个特定的表达式来匹配。尽管硬着头皮，也能在 C 层面上接入一个表达式引擎，但是纯 C 实现做不了非常灵活地自定义表达式里面的变量。
+
+举个例子，下面是 APISIX 用来匹配 GraphQL 请求的 route 配置：
+
+```json
+{
+        "methods": ["POST"],
```
Comment (Serendipity96) [inline _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md:138]: I think we need reformatting, there are too many spaces.

File: _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md (hunk 1 above)
Comment (yzeng25) [reply _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md:8]: 1. Apache APISIX
2. api7.ai

File: _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md (hunk 2 above)
Comment (yzeng25) [reply _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md:11]: 1. checked the whole article for APISIX --> Apache APISIX
2. add Apache APISIX Github link @line11

File: _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md (hunk 4 above)
Comment (yzeng25) [reply _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md:44]: fixed

File: _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md (hunk 6 above)
Comment (yzeng25) [reply _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md:119]: this is real, confirmed

File: _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md (hunk 5 above)
Comment (yzeng25) [reply _posts/blog/zh-CN/Why-Apache-APISIX-chose-Nginx-and-Lua.md:130]: fixed

**Review by Serendipity96 — APPROVED** [review]

**Pull request merged**